src/storage-client/src/sink.rs

```diff
@@ -352,11 +358,22 @@ pub struct ProgressRecord {
 
 /// Determines the latest progress record from the specified topic for the given
 /// key (e.g. akin to a sink's GlobalId).
+///
+/// For details about the uses of the two distinct consumers, see this
+/// function's implementation.
+///
+/// # Warning
+///
+/// To ensure that the returned timestamp is the latest timestamp, all
+/// transactions from previous epochs of this sink must be terminated. To
+/// accomplish this, start a new producer with the same transaction ID as the
+/// producer from the last epoch.
 async fn determine_latest_progress_record(
     name: String,
     progress_topic: String,
     progress_key: String,
-    progress_client: Arc<BaseConsumer<TunnelingClientContext<MzClientContext>>>,
+    progress_client_read_committed: Arc<BaseConsumer<TunnelingClientContext<MzClientContext>>>,
+    progress_client_read_uncommitted: Arc<BaseConsumer<TunnelingClientContext<MzClientContext>>>,
```

Comment:
There's likely some way of avoiding the `Arc` here but I couldn't find it. Suggestions welcome!

Reply:
You needed the `Arc` previously to make it through the retry loop, which needed a `clone` of the Kafka client per turn of the loop. And Kafka clients aren't clonable without wrapping them in an `Arc`. We could have pushed the retry logic into the sync function (swapping `retry_async` for `retry`) to avoid the clones, but the whole retry loop seemed unnecessary to me and so I removed it in https://github.com/MaterializeInc/materialize/pull/23438.